zkevm-circuits/src/mpt_circuit/helpers.rs

```diff
@@ -1,335 +1,1108 @@
+pub(crate) struct MainData<F> {
+    pub(crate) proof_type: Cell<F>,
+    pub(crate) is_below_account: Cell<F>,
+    pub(crate) address_rlc: Cell<F>,
+    pub(crate) root_prev: Cell<F>,
+    pub(crate) root: Cell<F>,
+}
+
+#[derive(Clone, Debug, Default)]
+pub(crate) struct MainDataWitness<F> {
+    pub(crate) proof_type: usize,
+    pub(crate) is_below_account: bool,
+    pub(crate) address_rlc: F,
+    pub(crate) root_prev: F,
+    pub(crate) root: F,
+}
+
+impl<F: Field> MainData<F> {
+    pub(crate) fn load(
+        description: &'static str,
+        cb: &mut ConstraintBuilder<F>,
+        memory: &MemoryBank<F>,
+        offset: Expression<F>,
+    ) -> Self {
+        let main_data = MainData {
+            proof_type: cb.query_cell(),
+            is_below_account: cb.query_cell(),
+            address_rlc: cb.query_cell(),
+            root_prev: cb.query_cell(),
+            root: cb.query_cell(),
+        };
+        circuit!([meta, cb], {
+            memory.load(
+                description,
+                cb,
+                offset,
+                &[
+                    main_data.proof_type.expr(),
+                    main_data.is_below_account.expr(),
+                    main_data.address_rlc.expr(),
+                    main_data.root_prev.expr(),
+                    main_data.root.expr(),
+                ],
+            );
+        });
+        main_data
+    }
+
+    pub(crate) fn store(
+        cb: &mut ConstraintBuilder<F>,
+        memory: &MemoryBank<F>,
+        values: [Expression<F>; 5],
+    ) {
+        memory.store(cb, &values);
+    }
+
+    pub(crate) fn witness_store(
+        _region: &mut Region<'_, F>,
+        offset: usize,
+        memory: &mut MemoryBank<F>,
+        proof_type: usize,
+        is_below_account: bool,
+        address_rlc: F,
+        root_prev: F,
+        root: F,
+    ) -> Result<(), Error> {
+        let values = [
+            proof_type.scalar(),
+            is_below_account.scalar(),
+            address_rlc,
+            root_prev,
+            root,
+        ];
+        memory.witness_store(offset, &values);
+
+        Ok(())
+    }
+
+    pub(crate) fn witness_load(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        memory: &MemoryBank<F>,
+        load_offset: usize,
+    ) -> Result<MainDataWitness<F>, Error> {
+        let values = memory.witness_load(load_offset);
+
+        self.proof_type.assign(region, offset, values[0])?;
+        self.is_below_account.assign(region, offset, values[1])?;
+        self.address_rlc.assign(region, offset, values[2])?;
+        self.root_prev.assign(region, offset, values[3])?;
+        self.root.assign(region, offset, values[4])?;
+
+        Ok(MainDataWitness {
+            proof_type: values[0].get_lower_32() as usize,
+            is_below_account: values[1] == 1.scalar(),
+            address_rlc: values[2],
+            root_prev: values[3],
+            root: values[4],
+        })
+    }
```

Comment:
I think I now understand the approach with `store/load` lookups. I very much like it! I would only perhaps change the names of the functions:
 - `witness_store` is intuitive - it stores the values in the array for later use
 - `witness_load` seems a bit less intuitive to me - what about `witness_assign`?
 - `store` adds an entry to the lookup table, so perhaps `add_to_lookup_table`?
 - `load` adds a lookup to be executed, so perhaps `add_lookup`?

But I trust your choices, you can leave it as it is if you disagree with the proposed names.

I have one question also - you added `key` to the table to enable adding lookups with `offset` different from 0, right? As it is now (`offset` always 0), the table could be without `key`?

Reply:
halo2 uses assign verb, so I would keep using that one, if it is about assigning a witness.

Reply:
Ok, I see. Thanks!